lib/Graph/Graph.cpp

```diff
@@ -565,27 +565,42 @@ FullyConnectedNode *Function::createFullyConnected(llvm::StringRef name,
 RowwiseQuantizedFullyConnectedNode *
 Function::createRowwiseQuantizedFullyConnected(llvm::StringRef name,
                                                NodeValue input, Constant *W,
-                                               Node *B, TypeRef outTy) {
+                                               Node *B, TypeRef outTy,
+                                               bool transposeWeight) {
   // Since W is constant, quantize it in compilation time.
   // The quantized data is in qWeights, the scale of each row is in scales,
   // and the offset of each row is in offsets.
   Constant *weights = llvm::cast<Constant>(W);
-  size_t numRows = W->getType()->dims()[0];
+  size_t numRows =
+      transposeWeight ? W->getType()->dims()[1] : W->getType()->dims()[0];
+  size_t numCols =
+      transposeWeight ? W->getType()->dims()[0] : W->getType()->dims()[1];
 
   // So far, if we want to create a storage with Int8QTy/Int16QTy,
   // it is assumed to be quantized data and the scale and offset should be
   // provided. But for rowwise quantization, the scales and offsets are stored
   // in vectors separately, we add the dummy scale and offset here.
-  auto *qWeights = getParent()->createConstant(ElemKind::Int8QTy, W->dims(),
-                                               0.0, 0, "weights.rwqfc");
+  auto *qWeights = getParent()->createConstant(
+      ElemKind::Int8QTy, {numRows, numCols}, 0.0, 0, "weights.rwqfc");
   auto *scales =
       getParent()->createConstant(ElemKind::FloatTy, {numRows}, "scales.rwqfc");
   auto *offsets = getParent()->createConstant(ElemKind::Int32ITy, {numRows},
                                               "offsets.rwqfc");
 
+  Tensor wt;
+  if (transposeWeight) {
+    // This happens when the RowwiseQuantizedFullyConnected node is converted
+    // from a quantized FullyConnected node in Glow's quantization procedure.
+    // Since in FC, the weights is stored as transposed (i.e. I * W + B), but in
+    // RowwiseQuantizedFullyConnected, the weights is stored as it is (i.e. I *
+    // W(T) + B).
```

Comment:
I'm kind of confused about this transposition business.  I assume we're doing this because Caffe2's RowwiseFC gives us transposed weights?  Can we handle transposition in the loader like we do for NCHW convolutions?

Reply:
So far, there is no `Caffe2 RowwiseFC op`. This op now is kind of internal and only for quantization.  In this quantization procedure, RowwiseQuantizedFullyConnected in created from FC, which has (I * W + B). But in the design of our RowwiseQuantizedFullyConnected node, it should be (I * W(T) + B) ( here the "rowwise" means the each row of "W").